### PR TITLE
@alloy => Migrate to WKWebView 

### PR DESCRIPTION
**Pull request by orta**
This still isn't 100% but it's getting there, and greens.

### CODE REVIEW

File: Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m [hunk 1]
```diff
@@ -38,38 +38,39 @@ - (BOOL)isContentStale;
...
+//    if (self.hasSuccessfullyLoadedLastRequest) {
+//        self.lastRequestLoadedAt = [NSDate date];
+//    }
+//}
+
+//- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error;
+//{
+//    [super webView:webView didFailLoadWithError:error];
+//    self.hasSuccessfullyLoadedLastRequest = NO;
+//}
```
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m]: We can't do this checking ( https://github.com/artsy/eigen/commit/afcd147f4ad5 ) WKWebView has its own cache we can't access.

Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m]: traded the URL check for a look at the progress

Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m]: What about `-[WKNavigationDelegate webView:decidePolicyForNavigationResponse:decisionHandler:]`, the `WKNavigationResponse` object has the `NSURLResponse` object.

**Comment by orta** [comment]
OK, this is feeling like it can only really get done with some pair programming, @jorystiefel want to do some with me tomorrow?

**Comment by jorystiefel** [comment]
@orta thought you'd never ask

**Comment by orta** [comment]
one more test, and then some on-device testing needing.

**Comment by alloy** [comment]
Groovy!

**Comment by orta** [comment]
This should probably green now. I will do some on-device testing.

**Comment by alloy** [comment]
And more rebasing

**Comment by orta** [comment]
OK, rebased, green'd locally and is ready for review/shipping.

**Comment by orta** [comment]
Fixes https://github.com/artsy/eigen/issues/288

File: Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m [hunk 2]
```diff
@@ -29,41 +25,55 @@ - (void)viewDidLoad
...
+}
+
+- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler
+{
+    decisionHandler([self shouldLoadNavigationAction:navigationAction]);
+}
+
+- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation;
+{
+    self.webView.scrollView.backgroundColor = [UIColor blackColor];
```
Comment (alloy) [inline Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m]: Can this not be set from `viewDidLoad` ?

File: Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m [hunk 3]
```diff
@@ -29,41 +25,55 @@ - (void)viewDidLoad
...
-- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error
+- (void)webView:(WKWebView *)webView didFailNavigation:(WKNavigation *)navigation withError:(NSError *)error
 {
     [self exitOnboarding];
 }
 
 - (void)exitOnboarding
 {
-    [self.personalizeDelegate dismissOnboardingWithVoidAnimation:YES];
+    [self.personalizeDelegate didSignUpAndLogin];
```
Comment (alloy) [inline Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m]: Is it correct that a user has signed-up and signed-in if this is called from the `webView:didFailNavigation:withError:` callback?

File: Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.h [hunk 4]
```diff
@@ -1,14 +1,17 @@
...
 
-- (instancetype)initWithURL:(NSURL *)url __attribute((objc_designated_initializer));
+- (NSURL *)currentURL;
+- (void)loadURL:(NSURL *)URL;
 
-// TSMiniBrowser setup doesn't happen if you don't use initWithURL.
-// Initializing without an NSURL also causes problems in tests.
-- (instancetype)init __attribute__((unavailable("Designated Initializer initWithURL: must be used.")));
+// Private
+- (WKNavigationActionPolicy)shouldLoadNavigationAction:(WKNavigationAction *)navigationAction;
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.h]: Is this for testing?

File: Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m [hunk 5]
```diff
@@ -34,9 +39,17 @@ - (void)viewDidLoad
...
+    webView.navigationDelegate = self;
+    [self.view addSubview:webView];
+
+    NSURLRequest *initialRequest = [NSURLRequest requestWithURL:self.initialURL];
+    [webView loadRequest:initialRequest];
+
+    webView.scrollView.delegate = self;
+    webView.scrollView.decelerationRate = UIScrollViewDecelerationRateNormal;
+
+    _webView = webView;
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m]: Use the dot, Luke!

File: Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m [hunk 6]
```diff
@@ -55,37 +68,39 @@ - (void)showAdminDetails:(UILongPressGestureRecognizer *)gesture
...
+//    NSString *webSession = [self.webView stringByEvaluatingJavaScriptFromString:@"sd.SESSION_ID"];
+//    [details appendFormat:@"Web Session: %@ \n", webSession];
+//
+//
+//    textView.text = details;
 #endif
 }
 
-- (void)setupConstraints
+- (void)viewWillLayoutSubviews
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m]: I think this should be `updateViewConstraints`

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m [hunk 7]
```diff
@@ -4,30 +4,17 @@
 #import "ARInternalShareValidator.h"
 #import "ARAppDelegate.h"
 
-@import TPDWeakProxy;
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: Also remove this pod from the Podfile.

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m [hunk 8]
```diff
@@ -52,7 +39,7 @@ - (instancetype)initWithURL:(NSURL *)url
     }
 
     if (![urlString isEqualToString:url.absoluteString]) {
-        ARActionLog(@"Rewriting %@ as %@", urlString, url.absoluteString);
+        NSLog(@"Rewriting %@ as %@", urlString, url.absoluteString);
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: `ARActionLog` isn’t working right anymore, is it?

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m [hunk 9]
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
...
-- (void)removeContentLoadStateTimer;
+- (void)viewDidLoad
 {
-    [self.contentLoadStateTimer invalidate];
-    self.contentLoadStateTimer = nil;
+    [super viewDidLoad];
+    [self showLoading];
+
+    // KVO on progress for when we can show the page
+    [self.webView addObserver:self forKeyPath:@"estimatedProgress" options:NSKeyValueObservingOptionNew & NSKeyValueObservingOptionOld context:ARProgressContext];
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: You can just pass `0` for the options here, right?

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m [hunk 10]
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
...
 }
 
-// A full reload, not just a webView.reload, which only refreshes the view without re-requesting data.
-- (void)userDidLoginOrSignUp
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
 {
-    [self reload];
+    if (context == ARProgressContext && [keyPath isEqualToString:@"estimatedProgress"]) {
+        [self.webView evaluateJavaScript:@"document.readyState == \"interactive\"" completionHandler:^(id response, NSError *error) {
+            if ([response respondsToSelector:@selector(boolValue)]) {
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: Could `response` be anything other than `nil` or `NSCFBoolean`? If not, just do `if ([response boolValue] && !self.loaded) {` ?

Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: And is the extra `!self.loaded` check really necessary? I guess sending the `-hideLoading` message when it’s already hidden isn’t actually bad, no?

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m [hunk 11]
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
...
 {
-    return [ARRouter requestForURL:url];
+    [self ar_presentIndeterminateLoadingIndicatorAnimated:YES];
 }
 
-- (void)reload;
+- (void)hideLoading
 {
-    [self.webView loadRequest:[self requestWithURL:self.currentURL]];
+    self.loaded = YES;
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: `self.loaded = YES` should only be set once the page has been fully loaded, as it is used as a check to see if the page was fully loaded when navigating back to the page. If setting it when hiding the progress indicator, it would be set to `YES` when `DOMContentLoaded` is reached but the page not fully loaded.

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m [hunk 12]
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
...
-}
-
-- (void)checkWebViewLoadingState;
-{
-    NSString *readyState = [self.webView stringByEvaluatingJavaScriptFromString:@"document.readyState"];
-    // DOMContentLoaded, which is once the webview is finished parsing but still loading sub-resources.
-    if ([readyState isEqualToString:@"interactive"]) {
-        [self removeContentLoadStateTimer];
-        [self hideLoading];
+    if (navigationAction.navigationType == WKNavigationTypeOther) {
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: Why is this needed, wouldn’t it just fall through to the final `return` statement?

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m [hunk 13]
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
...
-#pragma mark - UIScrollViewDelegate
+    NSLog(@"Martsy URL %@ - %@", URL, @(navigationAction.navigationType));
 
-- (UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView
-{
-    return nil;
-}
+    if (navigationAction.navigationType == WKNavigationTypeLinkActivated) {
+        if ([self.shareValidator isSocialSharingURL:URL]) {
+            ARWindow *window = ARAppDelegate.sharedInstance.window;
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: Use `self.webView.window` instead. Getting the right windows is going to get increasingly more important with multi-tasking etc, me thinks.

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m [hunk 14]
```diff
@@ -60,165 +47,132 @@ - (instancetype)initWithURL:(NSURL *)url
...
+            CGPoint lastTouchPointInView = [window convertPoint:window.lastTouchPoint toView:self.view];
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: Btw, I think you can just omit getting the window completely and instead use `-[UIView convertPoint:fromView:]` with `nil` as the `view` argument, which makes it cover from window coordinates.

File: Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m [hunk 15]
```diff
@@ -57,18 +57,18 @@ - (void)markRemoteNotificationsAsRead;
 
 #pragma mark - Overrides
 
-- (instancetype)initWithURL:(NSURL *)url;
+- (instancetype)initWithURL:(NSURL *)URL;
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m]: YES! I hate those lowercase `url` variables :D

File: Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m [hunk 16]
```diff
@@ -83,15 +83,9 @@ - (void)viewDidAppear:(BOOL)animated;
...
+- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation;
 {
-    [super webViewDidFinishLoad:webView];
-
-    NSCachedURLResponse *urlResponse = [[NSURLCache sharedURLCache] cachedResponseForRequest:webView.request];
-    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)urlResponse.response;
-    NSInteger statusCode = httpResponse.statusCode;
-    self.hasSuccessfullyLoadedLastRequest = statusCode >= 200 && statusCode < 300;
-
+    self.hasSuccessfullyLoadedLastRequest = webView.estimatedProgress == 1;
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m]: Is this really the same as the replaced code? The old code specifically checks for status codes, because e.g. a `5xx` still means that the webview has fully rendered the document, but a remote error occurred. If a remote error occurred we really want the webview to reload the next time.

File: Artsy_Tests/View_Controller_Tests/Web_Browsing/ARInternalMobileWebViewControllerTests.m [hunk 17]
```diff
@@ -7,10 +7,19 @@
...
+    [[[action stub] andReturnValue:OCMOCK_VALUE(type)] navigationType];
+    [[[action stub] andReturn:request] request];
+    return action;
+}
+
 
 @interface ARInternalMobileWebViewController (Testing)
 
 @property (nonatomic, strong) ARInternalShareValidator *shareValidator;
+- (NSURLRequest *)requestWithURL:(NSURL *)url;
```
Comment (alloy) [inline Artsy_Tests/View_Controller_Tests/Web_Browsing/ARInternalMobileWebViewControllerTests.m]: NoooooooOOOOooOoo, damned lowercase `url` variables come back to haunt me once again 😱

**Comment by alloy** [comment]
Maybe I missed it, but it appears that you are not stopping webview loading when the view disappears, which is what `TSMiniBrowser` used to do. With many of our web pages being resource heavy, it kinda makes sense to save the user’s bandwidth once they navigate away to another page.

This is also what the `loaded` ivar was for, so that the webview could be reloaded if it never finished loading the first time around and the user navigates back to the webview.

**Comment by alloy** [comment]
Intens change, I love it, finally gone with the ugly chrome and way better progress hooks! :clap:

File: Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m (hunk 2 above)
Comment (orta) [inline Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m]: No, it makes the full screen black, doing it here does it after it has content on it.

File: Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m (hunk 5 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m]: Hah, but it's a read only property, having it as read-write encourages more write-ness, but if this is the eigen standard - happy to oblige

Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m]: doing it this way means you can easily look for any setting specifically because it has to be done via the `_x = y` format

File: Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m (hunk 6 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m]: :+1:

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m (hunk 7 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: :tada:

File: Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m (hunk 2 above)
Comment (alloy) [inline Artsy/View_Controllers/Login_and_Onboarding/ARPersonalizeWebViewController.m]: Ah gotcha, so setting it to black earlier would make the whole view go black.

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m (hunk 10 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: I just didn't want to trust the javascript runtime - was being extra cautious. Happy to switch t something simpler.

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m (hunk 11 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: cool, thanks, will change logic

File: Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m (hunk 5 above)
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m]: But you _are_ writing to it, ain’t ya? :) Just make it read-write in the implementation only.

You can also look for it easily with `self.x`, so I don’t think that really matters.

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m (hunk 10 above)
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: If possible I’d prefer the simpler code, because otherwise the next time we need to change something around this it might seem as if there’s more things to consider than there really are.

File: Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m (hunk 16 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m]: this might have been a overzealous merge, will look

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m (hunk 12 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: Hrm, might be that the logic was shuffled and not re-evaulated at the end, will re-look.

**Comment by orta** [comment]
### TODO

- [ ] url - URL
- [ ] self.view.window
- [ ] status codes
- [ ] navigationAction.navigationType == WKNavigationTypeOther
- [ ] self.webView evaluateJavaScript

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m (hunk 14 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: we get the touch point _from_ the window, so we need the reference

Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: I think I'm keeping it this way, as it relies on behavior that can only come from the app delegate's uiwindow rather than any window

Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: That’s what the method does, though https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIView_Class/#//apple_ref/occ/instm/UIView/convertPoint:fromView:

Or I must be understanding the code wrong?

File: Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m (hunk 5 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m]: :P

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m (hunk 14 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: Yeah - we do  this: https://github.com/artsy/eigen/blob/master/Artsy/App/ARAppDelegate.m#L424-L435

Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: cause there's no connection between the touch event and the navigation call

Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: Oohhhh, I see :+1:

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m (hunk 9 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: not sure, so I opted for the most obvious

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m (hunk 10 above)
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: removing self.loaded, the `ar_removein...` checks if it needs to run.

**Comment by orta** [comment]
I've updated the PR - I've also removed the web view admin stuff. If I really care ,I'll go in and add something like https://github.com/Naituw/WBWebViewConsole

File: Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m [hunk 18]
```diff
@@ -60,79 +47,52 @@ - (instancetype)initWithURL:(NSURL *)url
...
-#pragma mark - UIViewController
-
-- (void)viewWillAppear:(BOOL)animated
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
 {
-    [super viewWillAppear:animated];
-
-    // As we initially show the loading, we don't want this to appear when you do a back or when a modal covers this view.
-    if (!self.loaded) {
-        [self reload];
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARInternalMobileWebViewController.m]: This reloading, when returning to the view and not fully loading previously, is still missing.

File: Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m [hunk 19]
```diff
@@ -83,12 +85,27 @@ - (void)viewDidAppear:(BOOL)animated;
...
 
-- (void)webViewDidFinishLoad:(UIWebView *)webView;
+// We want to be able to track the responses and the requests in
+// order to evaluate later the status code
+
+- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler
 {
-    [super webViewDidFinishLoad:webView];
+    [super webView:webView decidePolicyForNavigationAction:navigationAction decisionHandler:decisionHandler];
+    self.initialRequest = self.initialRequest ?: navigationAction.request;
```
Comment (alloy) [inline Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m]: The `initialRequest` needs to be set to `nil` when the webview is reloaded, just like `lastRequestLoadedAt` is set to `nil` from `loadURL:`. Otherwise there’s the chance that a second 500 would be considered ok and it won’t get reloaded.

Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m]: Legit :+1:

**Comment by orta** [comment]
force pushed changes requested this morning

**Comment by alloy** [comment]
@orta There’s still a bunch of CI errors about `shouldLoadNavigationAction:`, what gives?

**Comment by orta** [comment]
laziness, didn't run tests after my changes, will fix after swift at artsy

**Comment by alloy** [comment]
Groovy, baby!

**Comment by orta** [comment]
alright, all good again

**Comment by alloy** [comment]
Sweet! :dancers:
